subsys/cpp/cpp_guard.c

```diff
@@ -0,0 +1,68 @@
+/*
+ * Copyright (c) 2019 Erwin Rol <erwin@erwinrol.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <kernel.h>
+#include <sys/__assert.h>
+
+static K_MUTEX_DEFINE(guard_mutex);
+
+static bool init_has_run(uint64_t *guard_object)
+{
+	return (((uint8_t *)guard_object)[0] != 0);
+}
+
+static void set_init_has_run(uint64_t *guard_object)
+{
+	((uint8_t *)guard_object)[0] = 1;
+}
+
+/*
+ * This function is called before initialization takes place. If this
+ * function returns 1, either __cxa_guard_relase or __cxa_guard_abort
+ * must be called with the same argument. The first byte of the guard_object
+ * is not modified by this function.
+ *
+ * Returns 1 if the initialization in not yet complete, otherwise 0.
+ */
+int __cxa_guard_acquire(uint64_t *guard_object)
```

Comment:
Where is this API documented?  Is it specific to GCC?

Why is the guard object a `uint64_t`?  The GCC source uses `__guard` which it appears to typedef as `int`.

This needs to be documented somewhere so we have a clue for when other toolchains need to be supported.

Reply:
Clang seems to use uint64_t. After some more searching (see link) the ARM C++ ABI seems to say 32bit and the Itanium (and everybody else) seems to say 64bit. So this patch might need a `#ifdef` for arm (not sure what is with arm64) and the rest. 

But like I mentioned earlier, this is something that probably is better implemented in the toolchain (gcc, clang, what ever), but I have no real knowledge on what the state is, and if C++ is even a thing that the toolchain maintainers are interested in. 

https://android.googlesource.com/platform/bionic/+/c9a659c/libc/bionic/__cxa_guard.cpp